core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java

```diff
@@ -98,94 +156,287 @@ public GeoBoundingBoxQueryBuilder bottomRight(String geohash) {
+        bottomRight.resetLat(lat);
+        topLeft.resetLon(lon);
         return this;
     }
 
+    /**
+     * Adds bottom left corner.
+     *
+     * @param point bottom left corner of bounding box.
+     */
     public GeoBoundingBoxQueryBuilder bottomLeft(GeoPoint point) {
         return bottomLeft(point.lat(), point.lon());
     }
 
+    /**
+     * Adds bottom left corner.
+     *
+     * @param point bottom left corner of bounding box.
+     */
     public GeoBoundingBoxQueryBuilder bottomLeft(String geohash) {
         return bottomLeft(GeoHashUtils.decode(geohash));
     }
 
     /**
-     * Adds top right point.
+     * Adds top right corner.
      *
-     * @param lat The latitude
-     * @param lon The longitude
+     * @param lat latitude of top right corner.
+     * @param lon longitude of top right corner.
      */
     public GeoBoundingBoxQueryBuilder topRight(double lat, double lon) {
-        box[TOP] = lat;
-        box[RIGHT] = lon;
+        topLeft.resetLat(lat);
+        bottomRight.resetLon(lon);
         return this;
     }
 
+    /**
+     * Adds top right corner.
+     *
+     * @param point topRight corner of bounding box.
+     */
     public GeoBoundingBoxQueryBuilder topRight(GeoPoint point) {
         return topRight(point.lat(), point.lon());
     }
 
+    /**
+     * Adds top right corner.
+     *
+     * @param geohash topRight corner of bounding box.
+     */
     public GeoBoundingBoxQueryBuilder topRight(String geohash) {
         return topRight(GeoHashUtils.decode(geohash));
     }
 
+    /**
+     * Specify whether or not to try and fix broken/wrapping bounding boxes.
+     * If set to true, also enables ignoreMalformed thus disabling geo point
+     * validation altogether.
+     **/
     public GeoBoundingBoxQueryBuilder coerce(boolean coerce) {
+        if (coerce) {
+            this.ignoreMalformed = true;
+        }
         this.coerce = coerce;
         return this;
     }
 
+    /** Returns whether or not to try and fix broken/wrapping bounding boxes. */
+    public boolean coerce() {
+        return this.coerce;
+    }
+
+    /**
+     * Specify whether or not to ignore validation errors of bounding boxes.
+     * Can only be set if coerce set to false, otherwise calling this
+     * method has no effect.
+     **/
     public GeoBoundingBoxQueryBuilder ignoreMalformed(boolean ignoreMalformed) {
-        this.ignoreMalformed = ignoreMalformed;
+        if (coerce == false) {
+            this.ignoreMalformed = ignoreMalformed;
+        }
         return this;
     }
 
+    /** Returns whether or not to skip bounding box validation. */
+    public boolean ignoreMalformed() {
+        return ignoreMalformed;
+    }
+
     /**
      * Sets the type of executing of the geo bounding box. Can be either `memory` or `indexed`. Defaults
      * to `memory`.
      */
+    public GeoBoundingBoxQueryBuilder type(GeoExecType type) {
+        this.type = (type != null) ? type : DEFAULT_TYPE;
```

Comment:
I started changing the setters that turn null into default value back to throwing IAE, parser shouldn't allow null type value here either. But we can also do that later.